btree/src/storage.rs

```diff
@@ -1,50 +1,96 @@
 use memmap::MmapMut;
 use std::cell::UnsafeCell;
 
+use parking_lot::RwLock;
+use std::cmp::min;
+use std::collections::HashMap;
 use std::convert::TryInto;
 use std::fs::File;
 use std::io;
 use std::mem::ManuallyDrop;
 use std::sync::atomic::{AtomicU64, Ordering};
 
+/// Wrapper over an mmaped file with extension capabilities
+/// The underlying file is split into multiple pages, this makes way the pointers(slices) are never invalidated, at the expense of not being to able to
+/// operate on chunks spanning multiple pages as contiguous data.
+/// The API provided is mostly unsafe, as there is no aliasing checks, it's expected to be done at a higher level.
 pub struct MmapStorage {
-    mmap: ManuallyDrop<UnsafeCell<MmapMut>>,
+    pages: ManuallyDrop<PageTable>,
     file_len: AtomicU64,
-    allocated_size: u64,
+    allocated_size: AtomicU64,
     file: *mut File,
+    page_size: u64,
+}
+
+type PageId = u64;
+
+struct PageTable {
+    // XXX: A vector may be a decent choice too
+    // It may be possible to reduce locking here
+    lookup: RwLock<HashMap<PageId, Page>>,
```

Comment:
It may be possible to use [sled's pagetable](https://crates.io/crates/pagetable) too. Which sees to be (at first glance) a radix-tree linux style. I'm not sure of depending on `sled_sync` though. But it may be worth looking into.

Reply:
down the line probably fine, it doesn't seems too sled specific, but for now also fine without it